File: CRC.java
import java.util.*;

public class CRC {
    static String xor(String a, String b){
        String r=""; 
        for(int i=1;i<a.length();i++) r+= (a.charAt(i)==b.charAt(i))?'0':'1';
        return r;
    }
    static String divide(String msg, String poly){
        int n=poly.length();
        String tmp=msg.substring(0,n);
        for(int i=n;i<msg.length();i++){
            tmp = (tmp.charAt(0)=='1') ? xor(poly,tmp)+msg.charAt(i) : xor("0".repeat(n),tmp)+msg.charAt(i);
        }
        return (tmp.charAt(0)=='1') ? xor(poly,tmp) : xor("0".repeat(n),tmp);
    }
    public static void main(String[] a){
        Scanner s=new Scanner(System.in);
        System.out.print("Data: ");
        String data=s.next();
        System.out.print("Polynomial: ");
        String poly=s.next();

        int m=poly.length()-1;
        String senderData = data + "0".repeat(m);
        String crc = divide(senderData, poly);

        System.out.println("CRC: " + crc);
        String transmitted = data + crc;
        System.out.println("Transmitted: " + transmitted);

        // Receiver
        String remainder = divide(transmitted, poly);
        System.out.println("Receiver Remainder: " + remainder);
        System.out.println(remainder.equals("0".repeat(m)) ? "No Error" : "Error Detected");
    }
}

File: crc.c
#include <stdio.h>
#include <string.h>

void xor(char *a, char *b, int n) {
    for(int i = 1; i < n; i++)
        a[i] = (a[i] == b[i]) ? '0' : '1';
}

void crc(char *data, char *poly) {
    int m = strlen(poly) - 1;
    int n = strlen(data);
    char msg[200];
    strcpy(msg, data);
    for(int i = 0; i < m; i++) msg[n++] = '0';
    msg[n] = '\0';

    char tmp[200];
    strncpy(tmp, msg, strlen(poly));
    tmp[strlen(poly)] = '\0';

    for(int i = strlen(poly); i <= strlen(msg); i++) {
        if(tmp[0] == '1') xor(tmp, poly, strlen(poly));
        if(i < strlen(msg)) tmp[strlen(poly)-1] = msg[i];
    }

    printf("CRC: %s\n", tmp);
}

int main() {
    char data[100], poly[20];

    printf("Enter data in binary: ");
    scanf("%s", data);

    printf("Enter polynomial in binary: ");
    scanf("%s", poly);

    crc(data, poly);

    return 0;
}